engine/crates/runtime-local/src/hooks.rs

```diff
@@ -1,68 +1,102 @@
 use std::{collections::HashMap, sync::Arc};
 
-use runtime::hooks::{HeaderMap, HookError, HooksImpl, UserError};
+use runtime::{
+    error::GraphqlError,
+    hooks::{EdgeDefinition, HeaderMap, Hooks, NodeDefinition},
+};
+use tracing::instrument;
 pub use wasi_component_loader::{ComponentLoader, Config as HooksConfig};
 
-pub struct HooksWasi(ComponentLoader);
+pub struct HooksWasi(Option<ComponentLoader>);
 
 impl HooksWasi {
-    pub fn new(loader: ComponentLoader) -> Self {
+    pub fn new(loader: Option<ComponentLoader>) -> Self {
         Self(loader)
     }
 }
 
-#[async_trait::async_trait]
-impl HooksImpl for HooksWasi {
-    type Context = HashMap<String, String>;
+impl Hooks for HooksWasi {
+    type Context = Arc<HashMap<String, String>>;
 
-    async fn on_gateway_request(&self, headers: HeaderMap) -> Result<(Self::Context, HeaderMap), HookError> {
-        let context = Self::Context::new();
+    #[instrument(skip_all)]
+    async fn on_gateway_request(&self, headers: HeaderMap) -> Result<(Self::Context, HeaderMap), GraphqlError> {
+        let Some(ref loader) = self.0 else {
+            return Ok((Arc::new(HashMap::new()), headers));
+        };
+        let context = HashMap::new();
 
-        Ok(self
-            .0
+        loader
             .on_gateway_request(context, headers)
             .await
-            .map_err(to_local_error)?)
+            .map(|(ctx, headers)| (Arc::new(ctx), headers))
+            .map_err(|err| match err {
+                wasi_component_loader::Error::Internal(err) => {
+                    tracing::error!("on_gateway_request error: {err}");
+                    GraphqlError::internal_server_error()
+                }
+                wasi_component_loader::Error::User(err) => error_response_to_user_error(err),
+            })
     }
 
-    async fn authorized(
+    #[instrument(skip_all)]
+    async fn authorize_edge_pre_execution<'a>(
         &self,
-        context: Arc<Self::Context>,
-        input: Vec<String>,
-    ) -> Result<Vec<Option<UserError>>, HookError> {
-        let results = self
-            .0
-            .authorized(context, input)
+        context: &Self::Context,
+        definition: EdgeDefinition<'a>,
+        arguments: impl serde::Serialize + serde::de::Deserializer<'a> + Send,
+        _metadata: impl serde::Serialize + serde::de::Deserializer<'a> + Send,
+    ) -> Result<(), GraphqlError> {
+        let Some(ref loader) = self.0 else {
+            unreachable!("authorize_edge_pre_execution called without Wasm");
```

Comment:
We should fix this in the future. It will be a very common thing where the user adds an `@authorized` directive but fails to add the wasi side. Now we get a really ugly panic, and for sure this is not unreachable.